### PR TITLE
fix(router): sort possible routes by cost

### PR DESCRIPTION
**Pull request by btford**
/cc @rkirov

### CODE REVIEW

File: modules/angular2/src/router/instruction.js [hunk 1]
```diff
@@ -20,15 +20,18 @@ export class Instruction {
   matchedUrl:string;
   params:Map<string, string>;
   reuse:boolean;
+  cost:int;
```
Comment (rkirov) [inline modules/angular2/src/router/instruction.js]: the constructor param contains self cost, while member variable contains aggregate cost, i.e the sum of self cost and children's costs. Rename one of the two to differentiate.

File: modules/angular2/src/router/path_recognizer.js [hunk 2]
```diff
@@ -89,12 +93,13 @@ export class PathRecognizer {
   segments:List;
   regex:RegExp;
   handler:any;
+  cost;
```
Comment (rkirov) [inline modules/angular2/src/router/path_recognizer.js]: cost:number;

File: modules/angular2/src/router/path_recognizer.js [hunk 3]
```diff
@@ -61,21 +61,25 @@ function parsePathString(route:string):List {
...
+  var cost = 0;
 
   for (var i=0; i<segments.length; i++) {
     var segment = segments[i],
-      match;
+        match;
 
     if (isPresent(match = RegExpWrapper.firstMatch(paramMatcher, segment))) {
       ListWrapper.push(results, new DynamicSegment(match[1]));
+      cost += 100;
```
Comment (rkirov) [inline modules/angular2/src/router/path_recognizer.js]: I can't see how adding cost gives you different result than simply having the right order in the if-else-if branches.

File: modules/angular2/src/router/route_registry.js [hunk 4]
```diff
@@ -77,39 +77,48 @@ export class RouteRegistry {
...
-          component: parentComponent,
-          children: children,
-          matchedUrl: candidate['matchedUrl']
-        });
       }
     }
 
+    if (fullSolutions.length > 0) {
+      ListWrapper.sort(fullSolutions, (a, b) => {
+        return a.cost < b.cost ? -1 : 1;
```
Comment (rkirov) [inline modules/angular2/src/router/route_registry.js]: if it is just an expression you can have it on one line `(a, b) => a.cost < b.cost ? -1 : 1;`

**Comment by rkirov** [comment]
mostly nitpicks, and need to make tests pass

**Comment by btford** [comment]
Master seems to be broken – do you know what's up with that?

**Comment by rkirov** [comment]
Looks like broken by https://github.com/angular/angular/commit/111fa60a9352b2ccd064793494def9338b84d537

**Comment by vicb** [comment]
master has been fixed, could you please rebase ?

**Comment by btford** [comment]
Landed as 17392f663fc7d6ae0ddceaead9982696f4e838b1.

**Pull request merged**